### DISCUSSION TITLE
Freeze when pushing a scene points to a bug in TCA

### DISCUSSION THREAD

**Discussion by ghost** [pointfreeco/swift-composable-architecture#2263] [Q&A]
Hello :)

In our app, we found an issue where a push navigation freezes under a very strange set of conditions and our investigation points to TCA as being the source of the problem.

I'm attaching to this message a zipped project containing two versions of a toy app, one built using TCA and the other built using vanilla SwiftUI. The TCA version has the issue, the other doesn't.

Our app is very modular and each scene is a separate SPM module. We also have "utility" modules, one of which is for reusable views and other view goodies (`ViewUtilities`).

One of these reusable views is a custom web view that uses a `UIViewControllerRepresentable` to embed a `SFSafariViewController` inside a SwiftUI view, but is nothing out of the ordinary.

One of the scenes has in its view a declaration for `@Environment(\.dismiss)`, even though it's not actually used in the toy app (we do use it in our app).

The toy app is then a radically stripped-off version of our app, keeping only the absolute minimum needed to replicate the issue, with names changed.

With that preamble out of the way, the set of necessary conditions that cause the freeze is bizarre:

- `@Environment(\.dismiss) var dismiss` declared in the view, even without being used
- the presence of the `ViewUtilities` module in the package manifest, as a dependency of the module that has `@Environment(\.dismiss)` in its scene's view, even if the `ViewUtilities` module is not actually being imported by the module declaring that dependency
- the use of `SFSafariViewController` in the `ViewUtilities` module's custom web view.

Comment away the `dismiss` property in `Screen2`'s view, or comment away the inclusion of `ViewUtilities` as a dependency in the `Screen2` module, or use a view controller other than `SFSafariViewController` and the problem goes away in the TCA version of the toy app.

In the vanilla SwiftUI version of the toy app, however, having all 3 of the above conditions works fine and there is no freeze.

The project has a `notes.txt` file with a couple more details.

Thanks for looking it over.

https://github.com/pointfreeco/swift-composable-architecture/assets/91148917/a3b8b7fc-120e-4931-9f18-af32a58653ec

[Dismiss issue.zip](https://github.com/pointfreeco/swift-composable-architecture/files/11968270/Dismiss.issue.zip)

**Comment by ghost** [comment]
Ah, sorry, Brandon and Stephen, I mistakenly made this into a discussion rather than an issue report.

**Reply by mbrandonw** [reply]
Discussions are preferred over issues, especially when it is not obviously a problem with the library.

I don't have time to fully look into the issue, but the fact that removing things like environment variables fixes it leads me to believe it is a SwiftUI issue, and not anything TCA could possibly fix. Further, your vanilla example is not a true 1-to-1 comparison because it uses `@State`. It is well known that many things that work with `@State` mysteriously break when converted to `@ObservedObject`, so a more fair comparison would use observable objects.

**Reply by ghost** [reply]
Hi @mbrandonw. Thanks for the quick reply.

Fair point about the use of `@State` vs `@ObservedObject`. I'll write a second version of the vanilla SwiftUI app that uses `@ObservedObject` and see if I can replicated the issue.

**Comment by ghost** [comment]
You're correct, @mbrandonw. It's not a TCA bug. Here's the version of the toy app, using `@ObservedObject`, and the problem does exist. We'll do some more digging and then file a bug report with Apple.

Thanks again.

[SwiftUI - ObservableObject.zip](https://github.com/pointfreeco/swift-composable-architecture/files/11968841/SwiftUI.-.ObservableObject.zip)
